Source_Code/Server/ClueLessServer.java

```diff
@@ -402,52 +405,49 @@ public void processGameStartRequest(GameStartRequest gsr)
 
     public void processMoveRequest( MoveRequest mr )
     {
-       /* First check to see if the action request is from the player who holds the
-        * active turn status...
-        */
-       if( PlayerList.get( CurrentPlayerIndex ).PlayerName.equals(mr.PlayerName))
-       {
-
-          // TODO: update player location
-
-          // Announce the move to all players
-          sendToAllPlayers( new Notification("Player " + mr.PlayerName
-             + " moved " + mr.moveDirection) );
-
-
-          // Update player turn status, and move to next player
-          sendToPlayer( PlayerList.get( CurrentPlayerIndex ).PlayerName, new TurnUpdate( false ) );
-          nextPlayer();
-          sendToPlayer( PlayerList.get( CurrentPlayerIndex ).PlayerName, new TurnUpdate( true ) );
-       }
-
+        // processAction now checks if the sending player is allowed to move
+        if(turnTracker.CanMove)
+        {
+            // TODO: update player location. Only update the turn tracker on valid moves
+            turnTracker.move();
+            // Announce the move to all players
+            sendToAllPlayers(new Notification("Player " + mr.PlayerName
+                    + " moved " + mr.moveDirection));
+        }
+        else
+        {
+            sendToClient(mr.UniqueID, new Notification("You've already moved this turn!"));
+        }
     }
 
     public void processSuggestRequest( SuggestRequest sr )
     {
-        if( PlayerList.get( CurrentPlayerIndex ).PlayerName.equals(sr.PlayerName))
+        if(turnTracker.CanSuggest)
         {
+            turnTracker.suggest();
+            // processAction now checks if the sending player is allowed to move
             // TODO - Extend to give players choice of which card to use to refute
             // right now we are just sending back the first refutation
 
             sendToAllPlayers( new SuggestNotification( sr ) );
 
-            // TODO - update to start at the next player in the list
-            for (Player p : PlayerList )
+            for (int i = 1; i < PlayerList.size(); i++)  // Start at 1 to start with the next player. Don't check the current player
```

Comment:
Lets go with this implementation, but I ended up doing this with Collections.rotate (to get current player to first position) and then deleting the first player.
   
```
ArrayList<Player> refutingPlayers = PlayerList;
Collections.rotate(refutingPlayers, refutingPlayers.size() - CurrentPlayerIndex);
refutingPlayers.remove(0);
```